server/src/main/java/org/elasticsearch/gateway/PersistedClusterStateService.java

```diff
@@ -634,23 +672,48 @@ private void addMetaData(MetaData metaData) throws IOException {
             }
         }
 
-        public void commit(long currentTerm, long lastAcceptedVersion) {
+        public void commit(long currentTerm, long lastAcceptedVersion) throws IOException {
+            ensureOpen();
             try {
                 for (MetaDataIndexWriter metaDataIndexWriter : metaDataIndexWriters) {
-                    metaDataIndexWriter.commit(nodeId, currentTerm, lastAcceptedVersion);
+                    metaDataIndexWriter.prepareCommit(nodeId, currentTerm, lastAcceptedVersion);
+                }
+            } catch (Exception e) {
+                try {
+                    close();
+                } catch (Exception e2) {
+                    logger.warn("failed on closing cluster state writer", e2);
+                    e.addSuppressed(e2);
+                }
+                throw e;
+            } finally {
+                closeIfAnyIndexWriterHasTragedyOrIsClosed();
+            }
+            try {
+                for (MetaDataIndexWriter metaDataIndexWriter : metaDataIndexWriters) {
+                    metaDataIndexWriter.commit();
                 }
             } catch (IOException e) {
                 // The commit() call has similar semantics to a fsync(): although it's atomic, if it fails then we've no idea whether the
                 // data on disk is now the old version or the new version, and this is a disaster. It's safest to fail the whole node and
                 // retry from the beginning.
-                throw new IOError(e);
+                try {
+                    close();
+                } catch (Exception e2) {
+                    e.addSuppressed(e2);
+                    throw new IOError(e);
```

Comment:
Only throws an `IOError` if `close()` throws an exception?

Reply:
was also caught by the tests right away :)

Reply:
damn robots taking our jobs :robot: